cpp/src/arrow/util/iterator.h

```diff
@@ -106,44 +143,100 @@ class MapIterator : public Iterator<O> {
   std::unique_ptr<Iterator<I>> it_;
 };
 
-template <typename Fn, typename I, typename O = typename std::result_of<Fn(I)>::type>
+template <typename Fn,
+          typename I = typename std::remove_pointer<
+              internal::call_traits::argument_type<0, Fn>>::type,
+          typename O = typename std::result_of<Fn(I)>::type>
 std::unique_ptr<Iterator<O>> MakeMapIterator(Fn map, std::unique_ptr<Iterator<I>> it) {
   return std::unique_ptr<MapIterator<Fn, I, O>>(
       new MapIterator<Fn, I, O>(std::move(map), std::move(it)));
 }
 
-template <typename T>
-class EmptyIterator : public Iterator<T> {
+/// \brief Like MapIterator, but where the function can fail.
+template <
+    typename Fn,
+    typename I =
+        typename std::remove_pointer<internal::call_traits::argument_type<0, Fn>>::type,
+    typename O =
+        typename std::remove_pointer<internal::call_traits::argument_type<1, Fn>>::typ>
+class MaybeMapIterator : public Iterator<O> {
```

Comment:
Is there a point in having both `MapIterator` and `MaybeMapIterator`? AFAICS, these are not meant to be used for yielding very small datum, so shaving a couple CPU cycles doesn't seem important here.